synapse/events/__init__.py

```diff
@@ -199,15 +200,25 @@ def is_redacted(self):
         return self._dict.get("redacted", False)
 
 
-class EventBase(object):
+class EventBase(metaclass=abc.ABCMeta):
```

Comment:
Hadn't seen this before, my understanding is this allows using the `abc.abstractmethod` decorator, which is used to ensure all sub-classes implement that method (or property in this case)?

Reply:
yes, essentially. Specifically, it enforces `abc.abstractmethod` by raising an exception if an attempt is made to instantiate the class without having defined a concrete implementation of the method. Without the metaclass, `@abstractmethod` doesn't do a whole lot (except act as documentation).

(The underlying mechanic is that the metaclass arranges for `EventBase.__abstractmethods__` to be set. CPython has magic support for that attribute, and will refuse to instantiate a class with a non-empty `__abstractmethods__`: https://github.com/python/cpython/blob/3.8/Objects/typeobject.c if you're interested.)

Reply:
Thanks for the explanation!